libcontainer/nsenter/nsexec.c

```diff
@@ -890,6 +899,15 @@ void nsexec(void)
 				bail("failed to set uid to root");
 
 			/*
+			 * At this point we are now able to do the previously deferred
+			 * CLONE_NEWIPC (see the unshare(config.cloneflags) comment above).
+			 */
+			if (leftover_cloneflags) {
+				if (unshare(leftover_cloneflags) < 0)
+					bail("failed to unshare leftover namespaces");
```

Comment:
How `unshare(CLONE_NEWUSER | CLONE_NEWIPC) is different from `unshare(CLONE_NEWUSER)` followed by `unshare(CLONE_NEWIPC)`. IIUC, in both the cases new user namespace will be owner of IPC namespace.

Reply:
While I agree that that is true on paper, it isn't. There's a kernel bug (as discussed in #959) which results in SELinux labels (on `/dev/mqueue`) not being settable in a user namespace unless the IPC namespace was created after the user namespace was set up fully (including the `uid_map` and `setuid` stuff).

Reply:
That issue really does not give exact technical details. I am looking at kernel code, and owner of new ipc namespace is going to be newly created user namespace. If that's the case, then we should be able to just do.

unshare(CLONE_NEWUSER | CLONE_NEWIPC)
set_uid_gid_map
switch to root inside container
mount_dev_mqueue_and_label.

I am wondering why above workflow is not working and if it is a kernel bug which needs fixing.

Reply:
I tried reproducing problem on fedora 26 and I can reproduce it. I see following error message in journal.

`SELinux: mount invalid.  Same superblock, different security settings for (dev mqueue, type mqueue)`

Reply:
Looking at strace() output. Looks like first we tried selinux context mount of mqueue and that failed. I think due to above kernel message. And the we tried a non-context mount and tried lsexattr() and that lsetxattr failed. 

29777 mount("mqueue", "/root/runc-testing/rootfs/dev/mqueue", "mqueue", MS_NOSUID|MS_NODEV|MS_NOEXEC, "context=\"system_u:object_r:svirt"... <unfinished ...>
29777 <... mount resumed> )             = -1 EINVAL (Invalid argument)

29777 mount("mqueue", "/root/runc-testing/rootfs/dev/mqueue", "mqueue", MS_NOSUID|MS_NODEV|MS_NOEXEC, NULL <unfinished ...>
29777 <... mount resumed> )             = 0

29777 lsetxattr("/root/runc-testing/rootfs/dev/mqueue", "security.selinux", "system_u:object_r:svirt_sandbox_"..., 47, 0) = -1 EPERM (Operation not permitted)

Reply:
I think first failure happens because new IPC namespace was created, it creates an internal mount point of mqueue fs. And when user space tries to mount it with context=, it fails because it does not match the security context provided when kernel created internal mount point. IIUC, following kernel code is creating that internal mount point.

mq_init_ns() {
      ns->mq_mnt = kern_mount_data(&mqueue_fs_type, ns);
}

Reply:
More debugging. I think following check in selinux fails.

selinux_inode_setxattr() {
          if (!inode_owner_or_capable(inode)) {
          }
}

This inode should belong to /rootfs/dev/mqueue (one belonging mqueuefs). IIUC, calling thread is the one which is already inside container and has effective fsuid=1000. But interestingly inod->i_uid is 0. And 0 is not mapped inside container, so kuid_has_mapping() check fails too.

So question is, how did inode->i_uid is 0. IIUC, it is the fsuid=1000 which created this mqueue directory and mqueue mount point. I would think that inode->i_uid should been 1000 instead? What am I missing?

Reply:
@rhvgoyal I'll admit I never went through debugging this particular issue, but I agree with your conclusion. It looks like the reason why it's done that way is because of the `mq_open` (and related `mq_*` syscalls) that don't go through the VFS but need access to the `mqueue` mountpoint -- hence the mount is created each time a new IPC namespace is created.

More importantly, it looks like [we actually cannot ever set mount options if we are not in a user namespace](http://elixir.free-electrons.com/linux/latest/source/security/selinux/hooks.c#L840) -- `mqueue` is not whitelisted. So `xattr` is the only way to set them, and you need to pass [this `uid`-based check](http://elixir.free-electrons.com/linux/latest/source/security/selinux/hooks.c#L3170). And in order for the `uid`-based check to pass you need to have your user namespace set up properly (namely `kuid_has_mapping` needs to work).

Reply:
@rhvgoyal Ah, you saw the same thing as me (I was debugging in parallel :smile_cat:).

> This inode should belong to /rootfs/dev/mqueue (one belonging mqueuefs). IIUC, calling thread is the one which is already inside container and has effective fsuid=1000. But interestingly inode->i_uid is 0.

This will only be true in rootless containers. In a container that uses user namespaces but is running as root, the `unshare(...)` will happen with an `fsuid` of `0` (which will carry on to `mqueue`).

> And 0 is not mapped inside container, so kuid_has_mapping() check fails too.

Yeah. I think making `kuid_has_mapping` pass is the only way to be sure that things work out the way we want. But we could try something like `setfsuid(rootuid)` which should not affect anything else.

Reply:
Hmm..., I think it ties back to clone(CLONE_NEWUSER | CLONE_NEWIPC) call. This is called by process with fsuid=0. And kernel creates new ipc mount namespace, and also creates an internal mount point of mqueuefs. That in turn instantiates root inode and assigns i_uid from the calling thread. 
mqueue_get_inode() {
         inode->i_uid = current_fsuid();
        inode->i_gid = current_fsgid();
}

And that's how root inode of mqueuefs gets i_uid=0. When container process later tries to change the label, it fails as host uid 0 is not mapped inside container. 

So now I atleast understand the problem.

And Mrunal's patch is helping because unshare(CLONE_NEWIPC) is called by container process with fsuid=1000. That means mqeueufs root inode will get i_uid=1000 and container process will have the privileges to change selinux label.

Reply:
@rhvgoyal We can set our `fsuid` though (see my above comment). I'm working on another patch which try to do it that way.